text/0300-programmatic-toolkit.md

```diff
@@ -0,0 +1,745 @@
+   */
+  async notify(msg: IoMessage<T>);
+  
+  /**
+   * Notifies the host of a message that requires a response.
+   * 
+   * If the host does not return a response the suggested
+   * default response from the input message will be used.
+   */
+  async requestResponse(msg: IoRequest<T,U>): U | undefined;
+}
+```
+
+For every message, the toolkit calls `notify` or `requestResponse` is called respectively.
+
+The a generic `IoEmitter` is available to provide a familiar event emitter interface.
+This implementation will do nothing for messages and requests that don't have a registered listener, i.e. the default response will be used as required.
+
+```ts
+const io = new IoEmitter();
+io.on('deploy', async (msg: IoMessage) => { /* do something */ });
+io.once('deploy:confirm-security', async (msg: IoRequest<boolean>): boolean => {
+  // reject the deployment
+  return false;
+})
+```
+
+The `CommandLineIoHost` is available, implementing all the interaction behavior for the AWS CDK CLI [TODO: this might end up in a separate package]:
+
+```ts
+const io = new CdkCliHost({
+  isInteractive: true
+})
+```
+
+##### SdkProvider
+
+  [Alternative: Make this `CredentialsProvider` or `SdkConfigProvider` instead]
+
+The `AwsSdkProvider` defines how the toolkit interacts with AWS services.
+A default provider is available.
+You can provide a custom class implementing the `IAwsSdkProvider` interface.
+A standard `SdkProviderForCdk` is available that includes all features of the AWS CDK CLI [TODO: this might end up in a separate package].
+
+#### Cloud Executables
+
+A Cloud Executable is the primitive that defines a deployable AWS CDK application.
+AWS CDK apps might need to be synthesized multiple times with additional context values before they are ready.
+This concept is what the Cloud Executable represents.
+
+The most flexible way to obtain a Cloud Executable is from a function returning a `CloudAssembly`.
+Providing a function allows your implementation to integrate with the [Programmatic Toolkit]'s multi-pass requirement.
+The function will be called one or more times with the known context values of the current iteration.
+It must then use these values to synthesize and return a `CloudAssembly`.
+For all features (e.g. lookups) to work correctly, the `App()` must be instantiated with the received context values.
+Since it is not possible to update the context of an app, it must be created inside the function scope.
+
+A basic implementation would look like this:
+
+```ts
+const cx: ICloudExecutable = CloudExecutable.fromCloudAssemblyProducer(
+  async (context: Record<string, any>): cxapi.CloudAssembly => {
+    const app = new cdk.App({ context });
+    const stack = new cdk.Stack(app);
+    return app.synth();
+  }
+);
+```
+
+Existing applications, can also be used by simply passing an AWS CDK app directory (containing a `cdk.json` file):
+
+```ts
+const cx: ICloudExecutable = CloudExecutable.fromCdkApp("/path/to/cdk/app");
```

Comment:
There's also a mode where we do a `CloudExecutable` from an external executable without the use of a `cdk.json` (pass the info that's normally in `cdk.json` directly)

Reply:
Yes. I don't think that needs to go into the readme though.